src/components/tools/image-tool.tsx

```diff
@@ -201,6 +218,54 @@ export default class ImageToolComponent extends BaseComponent<IProps, IState> {
             : naturalHeight + kMarginsAndBorders;
   }
 
+  private handleBeginEditTitle = () => {
+    this.setState({ isEditingTitle: true });
+  };
+
+  private handleTitleChange = (title?: string) => {
+    title && this.getContent().setTitle(title);
+    this.setState({ isEditingTitle: false });
+  };
+
+  private renderTitleArea() {
+    return (
+      <div className="title-area-wrapper" key="title-area">
+        <div className="title-area">
+          {this.renderTitle()}
+        </div>
+      </div>
+    );
+  }
+
+  // Taken from ./hooks/use-measure-text.ts
+  // TODO: Need a unified solution here
+  private getMeasureText(font: string) {
+    const canvas = document.createElement("canvas");
+    const context = canvas.getContext("2d");
+    context && font && (context.font = font);
+    return (text: string) => {
+      const _context = canvas.getContext("2d");
+      return _context
+                ? Math.ceil(10 * _context.measureText(text).width) / 10
+                : 0;
+    };
+  }
```

Comment:
`canvas` and `cache` are both specific to each tile, which is especially odd for the cache. That means that if I name one tile 'g1' then 'g2' then 'g1', it won't recompute the size of 'g1'. But it will recompute the size of 'g1' for another tile if I name it that. Probably doesn't make sense to dedicate too much time to optimizing this given that it probably won't run all that much, but I'd be interested in knowing the best way to handle it.

Reply:
I ended up making some major changes here, which I think are improvements in consistency and performance for image tiles. The big challenge is that we have to write "hooks" for both functional and class components. In any case, I gave image tiles a cache, so they don't have to measure the title size on every render, and moved almost all of the logic into `use-measure-text.ts`. Several variables (a canvas, a cache, and a function) are now being defined in the `ImageTool` class component. I'm not sure this is the best way of handling this but it seems to work ok.